enclave_build/src/cache.rs

```diff
@@ -242,6 +242,145 @@ impl CacheManager {
+
+        // The manifest is checked, so now validate the layers from the manifest
+        self.validate_layers(&manifest_json)?;
+
+        // Extract the config digest from the manifest
+        let config_digest = Self::fetch_config_digest(manifest_json)?;
+
+        // Finally, check that the config is correctly cached
+        // This is done by applying a hash function on the config file contents and comparing the
+        // result with the config digest from the manifest
+        let config_str = self.fetch_config(&config_digest)?;
+
+        // Compare the two digests
+        if config_digest != format!("{:x}", sha2::Sha256::digest(config_str.as_bytes())) {
+            return Err(EnclaveBuildError::CacheMissError(
+                "Config content digest and manifest digest do not match".to_string(),
+            ));
+        }
+
+        Ok(())
+    }
+
+    /// Validates that the image layers are cached correctly by checking them with the layer descriptors
+    /// from the image manifest.
+    fn validate_layers(&self, manifest_obj: &Value) -> Result<()> {
+        let layers_path = self.root_path.join(CACHE_BLOBS_FOLDER);
+
+        // Try to get the layer list from the manifest JSON
+        let layers_vec = manifest_obj
+            .get("layers")
+            .ok_or_else(|| {
+                EnclaveBuildError::CacheMissError(
+                    "'layers' field missing from manifest JSON.".to_string(),
+                )
+            })?
+            .as_array()
+            .ok_or_else(|| {
+                EnclaveBuildError::CacheMissError("Manifest deserialize error.".to_string())
+            })?
+            .to_vec();
+
+        // Get the cached blobs as a HashMap mapping a layer digest to the corresponding layer file
+        let mut cached_blobs = HashMap::new();
+
+        fs::read_dir(layers_path)
+            .map_err(|err| {
+                EnclaveBuildError::CacheMissError(format!("Failed to get image layers: {:?}", err))
+            })?
+            .into_iter()
+            // Get only the valid directory entries that are valid files and return (name, file) pair
+            .filter_map(|entry| match entry {
+                Ok(dir_entry) => match File::open(dir_entry.path()) {
+                    Ok(file) => Some((dir_entry.file_name(), file)),
+                    Err(_) => None,
+                },
+                Err(_) => None,
+            })
+            // Map a layer digest to the layer file
+            // The 'cached_layers' hashmap will contain all layer files found in the cache for the current image
+            .for_each(|(name, file)| {
+                if let Ok(file_name) = name.into_string() {
+                    cached_blobs.insert(file_name, file);
+                }
+            });
+
+        // Iterate through each layer found in the image manifest and validate that it is stored in
+        // the cache by checking the digest
+        for layer_obj in layers_vec {
```

Comment:
This should be split into a separate function to keep this function simple and then you can also use a `map` or `for_each`.